tutorials/docs-17-implementing-samplers/index.qmd

```diff
@@ -0,0 +1,479 @@
+Turing.Inference.getparams(::Turing.Model, x::AbstractVector{<:Real}) = x
+```
+
+And with that, we're good to go!
+
+```{julia}
+# Our previous model defined as a Turing.jl model.
+@model mvnormal_model() = x ~ MvNormal([-5., 0., 5.], I)
+# Instantiate our model.
+turing_model = mvnormal_model()
+# Call `sample` but now we're passing in a Turing.jl `model` and wrapping
+# our `MALA` sampler in the `externalsampler` to tell Turing.jl that the sampler
+# expects something that implements LogDensityProblems.jl.
+chain = sample(turing_model, externalsampler(sampler), 10_000; progress=false)
+```
+
+Pretty neat, eh?
+
+### Models with constrained parameters
+
+One thing we've sort of glossed over in all of the above is that MALA, at least how we've implemented, requires $x$ to live in $\mathbb{R}^d$ for some $d > 0$. If some of the parameters were in fact constrained, e.g. we were working with a `Beta` distribution which has support on the interval $(0, 1)$, *not* on $\mathbb{R}^d$, we could easily end up outside of the valid range $(0, 1)$.
+
+```{julia}
+@model beta_model() = x ~ Beta(3, 3)
+turing_model = beta_model()
+chain = sample(turing_model, externalsampler(sampler), 10_000; progress=false)
+```
+
+Yep, that still works, but only because Turing.jl actually *transforms* the `turing_model` from constrained to unconstrained, so that the `sampler` provided to `externalsampler` is actually always working in unconstrained space! This is not always desirable, so we can turn this off:
+
+```{julia}
+chain = sample(turing_model, externalsampler(sampler; unconstrained=false), 10_000; progress=false)
+```
+
+The fun thing is that this still sort of works because
+
+```{julia}
+logpdf(Beta(3, 3), 10.0)
+```
+
+and so the samples that fall outside of the range are always rejected. But do notice how much worse all the diagnostics are, e.g. `ess_tail` is very poor compared to when we use `unconstrained=true`. Moreover, in more complex cases this won't just result in a "nice" `-Inf` log-density value, but instead will error:
+
+```{julia}
+@model function demo()
+    σ² ~ truncated(Normal(), lower=0)
+    # If we end up with negative values for `σ²`, the `Normal` will error.
+    x ~ Normal(0, σ²)
+end
+sample(demo(), externalsampler(sampler; unconstrained=false), 10_000; progress=false)
+```
+
+As expected, we run into a `DomainError` at some point, while if we set `unconstrained=true`, letting Turing.jl transform the model to a unconstrained form behind the scenes, everything works as expected:
+
+```{julia}
+sample(demo(), externalsampler(sampler; unconstrained=true), 10_000; progress=false)
+```
+
+Neat!
+
+Similarly, which automatic differentiation backend one should use can be specified through the `adtype` keyword argument too. For example, if we want to use [ReverseDiff.jl](https://github.com/JuliaDiff/ReverseDiff.jl) instead of the default [ReverseDiff.jl](https://github.com/JuliaDiff/ForwardDiff.jl):
```

Comment:
```suggestion
Similarly, which automatic differentiation backend one should use can be specified through the `adtype` keyword argument too. For example, if we want to use [ReverseDiff.jl](https://github.com/JuliaDiff/ReverseDiff.jl) instead of the default [ForwardDiff.jl](https://github.com/JuliaDiff/ForwardDiff.jl):
```